onnxruntime/python/onnxruntime_pybind_state.cc

```diff
@@ -324,6 +330,93 @@ const std::vector<std::string>& GetAvailableProviders() {
   return available_providers;
 }
 
+#ifdef USE_CUDA
+bool IsCudaDeviceIdValid(int id) {
+    
+    int num_devices = 0;
+    CUDA_CALL_THROW(cudaGetDeviceCount(&num_devices));
+
+    if (0 == num_devices)
+    {
+        printf("your system does not have a CUDA capable device.\n");
+        return false;
+    }
+
+    if (id < 0 || id >= num_devices)
+    {
+        printf("cuda_device=%d is invalid, must choose device ID between 0 and %d\n", id, num_devices - 1);
+        return false;
+    }
+
+    return true;
+}
+
+
+void UpdateCudaProviderOptions(onnxruntime::CudaProviderOptions& options, std::map<std::string, std::string> options_map) {
+    std::map<std::string, std::string>::iterator it;
+
+    it = options_map.find("device_id");
+    if (it != options_map.end()) {
+        OrtDevice::DeviceId device_id;
+        try {
+          device_id = std::stoi(it->second); 
+        }
+        catch (const std::invalid_argument& ia) {
+            std::cerr << "Invalid argument: " << ia.what() << '\n';
+            throw std::runtime_error("Please provide device id with integer.");
+            return;
+        }
+
+        if (!IsCudaDeviceIdValid(device_id)) {
+            throw std::runtime_error("Please provide available device id.");
+            return;
+        }
+        options.device_id = device_id; 
+        std::cout << "cuda device id is set to " << device_id << '\n';
+    }
+
+    it = options_map.find("cuda_mem_limit");
+    if (it != options_map.end()) {
+        size_t size;
+        try {
+          size = std::stoul(it->second, nullptr, 0);
+        }
+        catch (const std::invalid_argument& ia) {
+            std::cerr << "Invalid argument: " << ia.what() << '\n';
```

Comment:
you don't need to print error message when you throw std::runtime_error()
the error message will get bubbled up to python.
so it's not useful to print it twice.
you can test some these error paths to see what the user experiences.